irods/test/query_test.py

```diff
@@ -211,6 +215,78 @@ def test_query_for_data_object_with_utf8_name_python3(self):
             self.sess.data_objects.unregister(obj_path)
             os.remove(encoded_test_file)
 
+    class Issue_166_context:
+
+        def __init__(self, session, coll_path='small_test_collection', num_objects=8):
+            self.session = session
+            if '/' not in coll_path:
+                coll_path = '/{}/home/{}/{}'.format(self.session.zone, self.session.username, coll_path)
+            self.coll_path = coll_path
+            self.num_objects = num_objects
+            self.test_collection = None
+
+        def __enter__(self):
+            self.test_collection = helpers.make_test_collection( self.session, self.coll_path, obj_count=self.num_objects)
+            return lambda : self.session.query(Collection.name, DataObject.name). filter(Collection.name == self.test_collection.path)
+
+        def __exit__(self,*_):
+            if self.test_collection is not None:
+                self.test_collection.remove(recurse=True, force=True)
+
+    def test_query_first__166(self):
+
+        with self.Issue_166_context(self.sess) as buildQuery:
+            for i in self.Iterate_to_exhaust_statement_table:
+                buildQuery().first()
+
+    def test_query_one__166(self):
+
+        with self.Issue_166_context(self.sess, num_objects = self.More_than_one_batch) as buildQuery:
+
+            for i in self.Iterate_to_exhaust_statement_table:
+                query = buildQuery()
+                try:
+                    query.one()
+                except MultipleResultsFound:
+                    pass # irrelevant result
+
+    def test_query_one_iter__166(self):
+
+        with self.Issue_166_context(self.sess, num_objects = self.More_than_one_batch) as buildQuery:
+
+            for i in self.Iterate_to_exhaust_statement_table:
+
+                for x in buildQuery():
+                    break # single iteration
+
+    def test_query_paging_get_batches__166(self):
+
+        with self.Issue_166_context( self.sess, num_objects = self.More_than_one_batch) as buildQuery:
+
+            pages = [b for b in buildQuery().get_batches()]
+            self.assertTrue( len(pages) > 1 and len(pages[0]) < self.More_than_one_batch )
+
+            for i in self.Iterate_to_exhaust_statement_table:
+                for batch in buildQuery().get_batches():
+                    break # leave query after only one batch
```

Comment:
For my own understanding, this is where we are testing paging results, correct? Looks like we fetch `MAX_SQL_ROWS` results, then fetch the next page of results, then break. Would it be worth the time to verify the results of the paging? I only raise this concern because the `AUTO_CLOSE` option in the iRODS C API for GenQuery prevents paging results (i.e. You will still get results if you run the query again, but it will always be the first page of results). Not sure if that concern is valid for the PRC, but I thought I'd mention it anyway.

Reply:
Before my final push today, I did verify that all tests pass, and then also that they fail if I revert Chris(cjsmeele)'s commit.
Wrt the code cited above (sorry, I force-pushed so the code excerpt above no longer links a valid commit  ... )  each buildQuery() creates a new query.  So, I've really done a first query in which I fetch a page to determine the paging (PRC calls it "batch") size and assert it's less than the total number of results.  That's to make sure it's a valid test --  I want it to be an unfinished query, and that's only true if paging size < # query results. When I go into that final loop it's to test that the statement table is actually protected  by Chris's fix from being exhausted.

So, that's a full defense and why it adequately represents the purpose for which it was written.  Alan, if I understand your suggestion, I should add an another call to get_batches()  between  the first one, and before the loop starts, and make sure the first page isn't being repeated where ostensibly the second should be returned? Please tell me if I'm missing the point.

Reply:
Right, so, my concern can be illustrated by this scenario:

I run a query which nets 300 results. `MAX_SQL_ROWS` is hard-coded in the server at 256. Will the 257th result be the same as the 1st (i.e. repeated result page)? This is an observed behavior when using `AUTO_CLOSE` in the C API.

If we feel this situation is well-covered by tests already, then my concerns are absolved. :)